arch: arm: ARMv8-M address/address-range permissions checks (for secure firmware)

ioannisg opened the pull request:
An extension of #6845 for ARMv8-M systems with Security Extension.
Overlaps and blocked by #6845 

Update: #6845 is merged, so this PR contains only the added parts for Secure permissions

ioannisg committed:
arch: arm: Define and implement API for test target (Secure)

This commit defines and implements an internal ARMv8-M API
that allows the user to evaluate access permissions of memory
locations, based on the ARMv8-M Tests Target (TT, TTA)
instruction support (for Secure requests).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: implement cmse address range check (secure)

This commit contributes the implementation of the CMSE
address range permission checks based on the ARMv8-M
address range check intrinsics exlusicely for Secure state.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: convenience wrappers for C variable Non-Secure permissions

Declare and implement convenience wrappers to evaluate Non-Secure
read and write permissions of C variables using the cmse address
range intrinsics.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

pizi-nordic commented on arch/arm/include/cortex_m/cmse.h:
```diff
@@ -214,6 +214,122 @@ int arm_cmse_addr_range_readwrite_ok(u32_t addr, u32_t size, int force_npriv);
+ * @param addr The address for which the MPU region is requested
+ * @param p_region Output pointer to the location to store the MPU region
+ *
+ * @return non-zero if @ref region contains a valid MPU region, otherwise 0.
```
This is opposite to the Zephyr approach, where 0 return value 0 is a success.
What about returning region number (>= 0) on success and -ERRCODE on failure?

ioannisg replied:
We could probably do that, yes. We would need to define an "arbitrary" ERRCODE and use it throughout the use of this API.

ioannisg replied:
That has to be done in all similar functions, of course.

agross-linaro replied:
yeah let's conform to the usual zephyr return values.

ioannisg replied:
Good, I am going to update the PR with these changes. I am going to use -EINVAL to indicate invalid region.

pizi-nordic commented on arch/arm/core/cortex_m/cmse/arm_core_cmse.c:
```diff
@@ -70,3 +70,70 @@ int arm_cmse_addr_range_readwrite_ok(u32_t addr, u32_t size, int force_npriv)
 
+#if defined(CONFIG_ARM_SECURE_FIRMWARE)
+
+int arm_cmse_mpu_nonsecure_region_get(u32_t addr, u8_t *p_region)
```
Shouldn't we have similar function for checking secure region?

ioannisg replied:
We do already.  We have arm_cmse_mpu_region_get on line 10 which works for secure and for non-secure.

pizi-nordic replied:
Ok

pizi-nordic commented on arch/arm/core/cortex_m/cmse/arm_core_cmse.c:
```diff
@@ -70,3 +70,70 @@ int arm_cmse_addr_range_readwrite_ok(u32_t addr, u32_t size, int force_npriv)
+		addr_info.flags.nonsecure_read_ok;
+}
+
+int arm_cmse_addr_nonsecure_read_ok(u32_t addr, int force_npriv)
```
Shouldn't we have similar function for checking secure region?

ioannisg replied:
Yes, this is implemented above in this file (i.e. above this patch)

pizi-nordic replied:
Ok

pizi-nordic commented on arch/arm/core/cortex_m/cmse/arm_core_cmse.c:
```diff
@@ -70,3 +70,70 @@ int arm_cmse_addr_range_readwrite_ok(u32_t addr, u32_t size, int force_npriv)
+	return _arm_cmse_addr_nonsecure_read_write_ok(addr, force_npriv, 0);
+}
+
+int arm_cmse_addr_nonsecure_read_write_ok(u32_t addr, int force_npriv)
```
Shouldn't we have similar function for checking secure region?

pizi-nordic commented on arch/arm/core/cortex_m/cmse/arm_core_cmse.c:
```diff
@@ -131,9 +131,43 @@ int arm_cmse_addr_nonsecure_read_ok(u32_t addr, int force_npriv)
 	return _arm_cmse_addr_nonsecure_read_write_ok(addr, force_npriv, 1);
 }
 
+static int _arm_cmse_addr_range_nonsecure_read_write_ok(u32_t addr, u32_t size,
```
Also, could we have similar utils for secure domain? It will be hard to write tests/self-checks without them.

ioannisg committed:
arch: arm: refactor function to align with the adopted api

Refactor arm_cmse_mpu_region_get(.) function, so its signature,
documentation, and implementation align with the rest of functions
for obtaining MPU, SAU and IDAU region numbers via the TT instruction.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

agross-oss reviewed (approved):
aside from the return values, looks ok

ioannisg commented:
@pizi-nordic Update the PR as requested, and added 1 extra commit that refactors an earlier MPU_region_get function, so it aligns with the signature/implementation change you proposed.
